File: janMarkusMilan/assignments/twopointers/lc977/lc977.js
// Problem Statement #
// Given a sorted array, create a new array containing squares of all the numbers of the input array in the sorted order.

/*
First Thoughts (Brute Force):
  - Loop through the array
  - Square the input after each iteration (if we create a new array, then just insert the new values into the new array)
  - Sort from least to greatest

Second Thoughts (Two Pointer):
  - Make an array and fill with some arbitrary number
  - Define two pointers at beginning and end
  - Get the square at each value the pointers are on
  - On some condition, we pick the most suitable square and insert into the array
*/

// Brute Force Solution
// function make_squares(arr) {
//   for (let i = 0; i < arr.length; i++) {
//     arr[i] = arr[i] * arr[i];
//   }
//   arr.sort((a, b) => a - b);
//   return arr;
// }

function make_squares(arr) {
  // TODO: Write code here
  let res = Array(arr).fill(Infinity);
  let index = arr.length - 1;

  let leftPtr = 0;
  let rightPtr = arr.length - 1;
  while (leftPtr <= rightPtr) {
    let leftVal = arr[leftPtr] * arr[leftPtr];
    let rightVal = arr[rightPtr] * arr[rightPtr];

    if (leftVal <= rightVal) {
      res[index] = rightVal;
      rightPtr--;
    } else {
      res[index] = leftVal;
      leftPtr++;
    }

    index--;
  }

  return res;
}

console.log(`Squares: ${make_squares([-2, -1, 0, 2, 3])}`);
console.log(`Squares: ${make_squares([-3, -1, 0, 1, 2])}`);
console.log(`Squares: ${make_squares([-4, -1, 0, 3, 10])}`);

// Solution
// -----
// const n = arr.length;
// squares = Array(n).fill(0);
// let highestSquareIdx = n - 1;
// let left = 0,
//   right = n - 1;
// while (left <= right) {
//   let leftSquare = arr[left] * arr[left],
//     rightSquare = arr[right] * arr[right];
//   if (leftSquare > rightSquare) {
//     squares[highestSquareIdx] = leftSquare;
//     left += 1;
//   } else {
//     squares[highestSquareIdx] = rightSquare;
//     right -= 1;
//   }
//   highestSquareIdx -= 1;
// }

// return squares;

// -----

// Time complexity #
// The above algorithm’s time complexity will be O(N) as we are iterating the input array only once.

// Space complexity #
// The above algorithm’s space complexity will also be O(N); this space will be used for the output array.

File: mariannaFervenza/assignments/twopointers/lc977/lc977.py
# Problem Statement #
# Given a sorted array, create a new array containing squares of all the numbers of the input array in the sorted order.

# We have a sorted array and we need to return the square
# To do that we need to loop through each value of the array array


# squares[pow(index, 2)]  # pow(arr[left],2)


def make_squares(arr):

    # n = len(arr)
    # squares = [0 for x in range(n)]
    # highestSquareIdx = n - 1
    # left, right = 0, n - 1
    # while left <= right:
    #     leftSquare = arr[left] * arr[left]
    #     rightSquare = arr[right] * arr[right]
    #     if leftSquare > rightSquare:
    #         squares[highestSquareIdx] = leftSquare
    #         left += 1
    #     else:
    #         squares[highestSquareIdx] = rightSquare
    #         right -= 1
    #     highestSquareIdx -= 1

    # return squares

    length = len(arr)
    squares = [0 for x in range(length)]
    # TODO: Write your code here
    greatestIndex = length - 1
    left = 0  # first element in the array
    right = length - 1  # last element of the array

    while left <= right:  # loop to bring them closer and closer to each other while squaring them

        leftSquare = arr[left] * arr[left]  # pow(arr[left], 2)
        rightSquare = arr[right] * arr[right]  # pow(arr[right], 2)

        if leftSquare > rightSquare:

            squares[greatestIndex] = leftSquare
            left += 1

        else:

            squares[greatestIndex] = rightSquare
            right -= 1

        greatestIndex -= 1

        return squares


def main():

    print("Squares: " + str(make_squares([-2, -1, 0, 2, 3])))
    print("Squares: " + str(make_squares([-3, -1, 0, 1, 2])))


main()


# Solution
# -----
#   n = len(arr)
#   squares = [0 for x in range(n)]
#   highestSquareIdx = n - 1
#   left, right = 0, n - 1
#   while left <= right:
#     leftSquare = arr[left] * arr[left]
#     rightSquare = arr[right] * arr[right]
#     if leftSquare > rightSquare:
#       squares[highestSquareIdx] = leftSquare
#       left += 1
#     else:
#       squares[highestSquareIdx] = rightSquare
#       right -= 1
#     highestSquareIdx -= 1

#   return squares

#   -----

#   Time complexity #
# The above algorithm’s time complexity will be O(N) as we are iterating the input array only once.

# Space complexity #
# The above algorithm’s space complexity will also be O(N); this space will be used for the output array.
